Loop/Managers/Alerts/AlertStore.swift

```diff
@@ -0,0 +1,149 @@
+
+        let storeDescription = NSPersistentStoreDescription()
+        if let storageFileURL = storageFileURL {
+            storeDescription.url = storageFileURL
+            print("AlertStore: \(storageFileURL)")
+        } else {
+            storeDescription.type = NSInMemoryStoreType
+        }
+        persistentContainer = NSPersistentContainer(name: "AlertStore")
+        persistentContainer.persistentStoreDescriptions = [storeDescription]
+        persistentContainer.loadPersistentStores { description, error in
+            if let error = error {
+                fatalError("Unable to load persistent stores: \(error)")
+            }
+        }
+        managedObjectContext.persistentStoreCoordinator = persistentContainer.persistentStoreCoordinator
+    }
+}
+
+// MARK: Alert Recording
+
+extension AlertStore {
+    
+    public func recordIssued(alert: DeviceAlert, at timestamp: Date = Date(), completion: ((Result<Void, Error>) -> Void)? = nil) {
+        dataAccessQueue.async {
+            self.managedObjectContext.perform {
+                _ = StoredAlert(from: alert, context: self.managedObjectContext, issuedTimestamp: timestamp)
+                do {
+                    try self.managedObjectContext.save()
+                    self.log.default("Recorded alert: %{public}@", alert.identifier.value)
+                    completion?(.success)
+                } catch {
+                    self.log.error("Could not store alert: %{public}@", String(describing: error))
+                    completion?(.failure(error))
+                }
+            }
+        }
+    }
+    
+    private func lookupLatest(identifier: DeviceAlert.Identifier, completion: @escaping (Result<StoredAlert?, Error>) -> Void) {
+        managedObjectContext.perform {
+            let predicate = NSPredicate(format: "identifier = %@", identifier.value)
+            do {
+                let fetchRequest: NSFetchRequest<StoredAlert> = StoredAlert.fetchRequest()
+                fetchRequest.predicate = predicate
+                fetchRequest.sortDescriptors = [ NSSortDescriptor(key: "modificationCounter", ascending: true) ]
+                let result = try self.managedObjectContext.fetch(fetchRequest)
+                completion(.success(result.last))
+            } catch {
+                completion(.failure(error))
+            }
+        }
+    }
+    
+    private func recordUpdateOfLatest(of identifier: DeviceAlert.Identifier, at timestamp: Date = Date(),
+                                      with block: @escaping (StoredAlert) -> Void,
+                                      completion: ((Result<Void, Error>) -> Void)?) {
+        dataAccessQueue.async {
+            self.managedObjectContext.perform {
+                self.lookupLatest(identifier: identifier) {
+                    switch $0 {
+                    case .success(let object):
+                        if let object = object {
+                            block(object)
```

Comment:
Any need to have the block return an error?

Alternatively, consider passing the *optional* object to the block and having it return an error so the block can decide if a record not found is an error or not.

Reply:
These are good suggestions, if this was a more public function.  This is a private implementation detail, though, of `recordRetraction` and `recordAcknowledgement`, and my goal was to minimize code duplication in those functions.  As such, the optional check and return (or "completion") of an error would also be duplicated.  
I'm not strongly opposed to your suggestion, I just don't really see a need for it.  Thoughts?

Reply:
Works for me.